Mathlib/RingTheory/Radical.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2024 Jineon Back and Seewoo Lee. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jineon Baek, Seewoo Lee
+-/
+import Mathlib.RingTheory.UniqueFactorizationDomain
+
+/-!
+# Radical of an element of a unique factorization normalization monoid
+
+This file defines a radical `rad(a)` of an element `a` of a unique factorization normalization
+monoid, which is defined as a product of normalized prime factors of `a` without duplication.
+This is different from the radical of ideal.
+
+## Main declarations
+
+- `radical_associated_eq`: If `a` and `b` are associates, i.e. `a * u = b` for some unit `u`,
+  then `rad(a) = rad(b)`.
+- `radical_unit_hMul`: Multiplying unit does not change the radical.
+- `radical_dvd_self`: `rad(a)` divides `a`.
+- `radical_pow`: `rad(a ^ n) = rad(a)` for any `n ≥ 1`
+- `radical_prime`: Radical of a prime element is equal to its normalization
+
+## TODO
+
+- Make a comparison with `Ideal.radical`. Especially, for principal ideal, `rad((a)) = (rad(a))`,
+  where the first (resp. second) `rad` is radical of an ideal (resp. element).
+- Prove `rad(rad(a)) = rad(a)`.
+-/
+
+noncomputable section
+
+open scoped Classical
+
+open UniqueFactorizationMonoid
+
+variable {k : Type _} [Field k]
+variable {α : Type _} [CancelCommMonoidWithZero α] [NormalizationMonoid α]
+  [UniqueFactorizationMonoid α]
+
+/-- Prime factors of a polynomial `a` are monic factors of `a` without duplication. -/
+def primeFactors (a : α) : Finset α :=
+  (normalizedFactors a).toFinset
+
+/-- Radical of a polynomial `a` is a product of prime factors of `a`. -/
+def radical (a : α) : α :=
+  (primeFactors a).prod id
+
+theorem radical_zero_eq_one : radical (0 : α) = 1 := by
+  rw [radical, primeFactors, normalizedFactors_zero, Multiset.toFinset_zero, Finset.prod_empty]
+
+theorem radical_one_eq_one : radical (1 : α) = 1 := by
+  rw [radical, primeFactors, normalizedFactors_one, Multiset.toFinset_zero, Finset.prod_empty]
+
+theorem radical_associated_eq {a b : α} (h : Associated a b) : radical a = radical b := by
+  rcases iff_iff_and_or_not_and_not.mp h.eq_zero_iff with (⟨rfl, rfl⟩ | ⟨ha, hb⟩)
+  · rfl
+  · simp_rw [radical, primeFactors]
+    rw [(associated_iff_normalizedFactors_eq_normalizedFactors ha hb).mp h]
```

Comment:
```suggestion
theorem radical_eq_of_associated {a b : α} (h : Associated a b) : radical a = radical b := by
  rcases iff_iff_and_or_not_and_not.mp h.eq_zero_iff with (⟨rfl, rfl⟩ | ⟨ha, hb⟩)
  · rfl
  · simp_rw [radical, primeFactors]
    rw [(associated_iff_normalizedFactors_eq_normalizedFactors ha hb).mp h]
    
alias Associated.radical_eq := radical_eq_of_associated
```

Reply:
Also, if you change the name as suggested, please also update the name in the documentation at the top of the file.